File: plugins/maestro/skills/maestro/scripts/maestro_stats.py
#!/usr/bin/env python3
"""Print Maestro's selected bundle folder and Known Bundles table."""

from __future__ import annotations

import argparse
import pathlib

try:
    import maestro_fetch as fetch
except ImportError as exc:  # pragma: no cover
    raise SystemExit(f"cannot import maestro_fetch.py from script directory: {exc}")


def known_rows(skill_file: pathlib.Path):
    text = fetch.read_text(skill_file, limit=1_000_000)
    return fetch.parse_existing(fetch.known_bundles_block(text))


def main() -> int:
    parser = argparse.ArgumentParser(description="Print Maestro Known Bundles table.")
    parser.add_argument(
        "--skill-file",
        default=str(pathlib.Path(__file__).resolve().parents[1] / "SKILL.md"),
        help="Maestro SKILL.md path. Defaults to this skill's SKILL.md.",
    )
    args = parser.parse_args()

    skill_file = pathlib.Path(args.skill_file).expanduser().resolve()
    if not skill_file.is_file():
        raise SystemExit(f"skill file not readable: {skill_file}")

    bundle_root = fetch.load_saved_bundle_root(skill_file)
    print("## Known Bundles")
    print()
    print(fetch.render_bundle_root(bundle_root))
    print()
    print(fetch.render_index(known_rows(skill_file)))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: maestro/scripts/maestro_stats.js
#!/usr/bin/env node
"use strict";
/**
 * Print Maestro's selected bundle folder and Known Bundles table.
 * Node.js port of maestro_stats.py.
 */

const fs   = require("fs");
const path = require("path");
const fetch = require("./maestro_fetch.js");

function knownRows(skillFile) {
  const text = fetch.readText(skillFile, 1000000);
  return fetch.parseExisting(fetch.knownBundlesBlock(text));
}

function main() {
  const args = process.argv.slice(2);
  let skillFileArg = path.resolve(path.join(__dirname, "..", "SKILL.md"));

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--skill-file" && args[i + 1]) { skillFileArg = args[++i]; }
  }

  const skillFile = path.resolve(skillFileArg);
  if (!fs.existsSync(skillFile)) {
    console.error(`skill file not readable: ${skillFile}`);
    process.exit(1);
  }

  const bundleRoot = fetch.loadSavedBundleRoot(skillFile);
  console.log("## Known Bundles");
  console.log();
  console.log(fetch.renderBundleRoot(bundleRoot));
  console.log();
  console.log(fetch.renderIndex(knownRows(skillFile)));
}

if (require.main === module) {
  main();
}
